festim/exports/txt_export.py

```diff
@@ -44,90 +62,157 @@ def filename(self, value):
+
         if self.times is None:
             return True
         for time in self.times:
             if np.isclose(time, current_time, atol=0):
                 return True
         return False
 
-    def when_is_next_time(self, current_time):
-        if self.times is None:
-            return None
-        for time in self.times:
-            if current_time < time:
-                return time
-        return None
+    def is_last(self, current_time, final_time):
+        """
+        Checks if the current simulation step equals to the last export time.
+        based on the final simulation time, TXTExport.times, and the current time
+
+        Args:
+            current_time (float): the current simulation time.
+            final_time (float, None): the final simulation time.
+
+        Returns:
+            bool: True if simulation is steady (final_time is None), if TXTExport.times
+                are not provided and the current time equals to the final time, or if
+                TXTExport.times are provided and the current time equals to the last time in
+                TXTExport.times, else False.
+        """
+
+        if final_time is None:
+            # write if steady
+            return True
+        elif self.times is None:
+            if np.isclose(current_time, final_time, atol=0):
+                # write at final time if exports at each timestep
+                return True
+        else:
+            if np.isclose(current_time, self.times[-1], atol=0):
+                # write at final time if exports at specific times
+                return True
+        return False
+
+    def initialise(self, mesh, project_to_DG=False, materials=None):
+        """
+        Initialises TXTExport. Depending on the project_to_DG flag, defines a function space (DG1 or CG1)
+        for projection of the exported field. After that, an unsorted array of mesh vertices is created for export.
+        The array is then used to obtain indices of sorted elements for the data export.
+
+        .. note::
+            If DG1 is used, the duplicated vertices in the array are filtered except those near interfaces,
+            The interfaces are defined by ``material.borders`` in the ``Materials`` list.
+
+        Args:
+            mesh (fenics.Mesh): the mesh.
+            project_to_DG (bool): if True, the exported field is projected to a DG1 function space.
+                Defaults to False.
+            materials (festim.Materials): the materials. Defaults to None.
+        """
+
+        if project_to_DG:
+            self.V = f.FunctionSpace(mesh, "DG", 1)
+        else:
+            self.V = f.FunctionSpace(mesh, "CG", 1)
+
+        x = f.interpolate(f.Expression("x[0]", degree=1), self.V)
+        x_column = np.transpose([x.vector()[:]])
+
+        # if project_to_DG is True, get indices of duplicates near interfaces
+        # and indices of the first elements from a pair of duplicates otherwise
+        if project_to_DG:
```

Comment:
I thought a bit about filtering, maybe we should allow users to control whether the data will be filtered or not?